src/time.c

```diff
@@ -101,3 +101,21 @@ void gps_time_match_weeks(gps_time_t *t, const gps_time_t *ref)
   else if (dt < -WEEK_SECS / 2)
     t->wn++;
 }
+
+/** Adjust the week number of t to correctly refelct the current week cycle.
+ *
+ * Assumes the current week number cannot be earlier than the reference WN. So
+ * will return the correct WN for at most 20 years after the reference WN.
+ *
+ * TODO the reference value should be set from the build time, as well as being
+ * able to be changed via a flash setting.
+ */
+void gps_adjust_week_cycle(gps_time_t *t, u16 wn_ref) {
```

Comment:
Also as mentioned I think we should take in only the uncorrected week number and adopt the convention that gps_time_t only ever contains the corrected / absolute week number.